src/utils/tst/TimerQueue.cpp

```diff
@@ -733,4 +740,49 @@ TEST_F(TimerQueueFunctionalityTest, updateTimerPeriodFunctionalityTest)
     EXPECT_TRUE(newInvocationCount > oldInvocationCount && oldInvocationCount + 1 == newInvocationCount);
 
     EXPECT_EQ(STATUS_SUCCESS, timerQueueFree(&handle));
-}
\ No newline at end of file
+}
+
+TEST_F(TimerQueueFunctionalityTest, shutdownTimerQueueTest)
+{
+    TIMER_QUEUE_HANDLE handle = INVALID_TIMER_QUEUE_HANDLE_VALUE;
+    UINT32 timerId;
+    UINT64 curTime;
+
+    // Make sure we don't check for the timer in the test callback
+    checkTimerId = FALSE;
+
+    // Create a valid timer queue
+    EXPECT_EQ(STATUS_SUCCESS, timerQueueCreate(&handle));
+
+    EXPECT_EQ(STATUS_SUCCESS, timerQueueAddTimer(handle, 0, 200 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND, testTimerCallback, (UINT64) this, &timerId));
+
+    THREAD_SLEEP(50 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+
+    EXPECT_EQ(1, ATOMIC_LOAD(&invokeCount));
+
+    // Call a shutdown with invalid param
+    EXPECT_NE(STATUS_SUCCESS, timerQueueShutdown(INVALID_TIMER_QUEUE_HANDLE_VALUE));
+
+    // Wait for some time for the timer to fire again
+    THREAD_SLEEP(250 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+
+    EXPECT_EQ(2, ATOMIC_LOAD(&invokeCount));
+
+    // Make it sleep next time it fires
+    ATOMIC_STORE(&sleepFor, 3 * HUNDREDS_OF_NANOS_IN_A_SECOND);
+
+    // Wait for the next firing
+    THREAD_SLEEP(250 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+
+    // We should now block until the timer thread returns
+    curTime = GETTIME();
+    EXPECT_EQ(STATUS_SUCCESS, timerQueueShutdown(handle));
+
+    // Should have at least 3 seconds passed
+    EXPECT_LE(curTime, GETTIME());
```

Comment:
This will always evaluate to true even in the bad path. I think we should add an acceptable to range to pass. Probably, [`curTime`, `curTime + 3 seconds + 300 ms (for the noise)`]?

Reply:
Oh, of course.

Reply:
I am also thinking about the case where an API can be called after the shutdown. Need to add a check and a few tests.

Will be back with a quick change